### PR TITLE
Use CS Core on supported platforms for USB cameras

### PR DESCRIPTION
**Pull request by ThadHouse**
Can be more reliable then OpenCV, and handles disconnects properly

Depends on #955 to build on azure.

### CODE REVIEW

**Comment by JLLeitschuh** [comment]
Status on this?

File: core/src/main/java/edu/wpi/grip/core/sources/CSHttpCameraFrameGrabber.java [hunk 1]
```diff
@@ -1,97 +1,101 @@
...
+  private JavaCvSink javaCvSink;
+  private HttpCamera httpCamera;
+
+  public CSHttpCameraFrameGrabber(String urlstr, int readTimeout, TimeUnit unit) {
+    super();
+    this.url = checkNotNull(urlstr, "urlstr");
+    this.readTimeout = TimeUnit.MILLISECONDS.convert(readTimeout, unit) / 1000.0;
+  }
+
+  public static void tryLoad() throws Exception {
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/CSHttpCameraFrameGrabber.java:38]: Where is this called?

File: core/src/main/java/edu/wpi/grip/core/sources/CSHttpCameraFrameGrabber.java [hunk 2]
```diff
@@ -1,97 +1,101 @@
...
+      javaCvSink.close();
+    }
+
+    if (httpCamera != null) {
+      httpCamera.close();
+    }
+
+    if (decoded != null) {
+      decoded.close();
+    }
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/CSHttpCameraFrameGrabber.java:74]: Do we want to try to recover and close the other stuff if earlier things fail to close?

File: core/src/main/java/edu/wpi/grip/core/sources/CSUsbCameraFrameGrabber.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        decoded = new Mat();
+      }
+      long frameTime = javaCvSink.grabFrame(decoded, readTimeout);
+      if (frameTime > 0) {
+        return converter.convert(decoded);
+      } else {
+        throw new IOException("Frame not read: " + frameTime);
+      }
+    } catch (IOException e) {
+      throw new Exception(e.getMessage(), e);
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/sources/CSUsbCameraFrameGrabber.java:92]: Why throw the raw 'Exception' type here? Seems like less information here?

File: core/src/main/java/edu/wpi/grip/core/sources/CSHttpCameraFrameGrabber.java (hunk 1 above)
Comment (ThadHouse) [reply core/src/main/java/edu/wpi/grip/core/sources/CSHttpCameraFrameGrabber.java:38]: I don't know. It was in the existing code that this was renamed from.

File: core/src/main/java/edu/wpi/grip/core/sources/CSUsbCameraFrameGrabber.java (hunk 3 above)
Comment (ThadHouse) [reply core/src/main/java/edu/wpi/grip/core/sources/CSUsbCameraFrameGrabber.java:92]: Was just copying what the original did. Whoever wrote it originally must have had a reason.

File: core/src/main/java/edu/wpi/grip/core/sources/CSHttpCameraFrameGrabber.java (hunk 2 above)
Comment (ThadHouse) [reply core/src/main/java/edu/wpi/grip/core/sources/CSHttpCameraFrameGrabber.java:74]: Was just copying what was already existing.

**Comment by ThadHouse** [comment]
@JLLeitschuh Everything you mentioned has been in the repo for 4 years at least. It originally came from the IP frame grabber.

https://github.com/WPIRoboticsProjects/GRIP/blob/master/core/src/main/java/edu/wpi/grip/core/sources/IPCameraFrameGrabber.java

**Comment by JLLeitschuh** [comment]
@ThadHouse is this finished and ready to be merged?

**Comment by ThadHouse** [comment]
Yes

**Pull request merged**